Add setting to enable string escaping in Logstash pipeline config

jordansissel opened the pull request:
This PR adds a new boolean setting `config.support_escapes` which defaults to false (the historical behavior). When set to true, the following escapes are handled:

* backslash doublequote -> doublequote
* backslash quote -> quote
* backslash n -> newline (ascii 10)
* backslash r -> carriage return (ascii 13)
* backslash backslash -> backslash
* backslash t -> tab (ascii 9)

This will solve #1645.

suyograo commented on config/logstash.yml:
```diff
@@ -85,6 +85,11 @@
+# When enabled, process escaped characters such as \n and \" in strings in the
+# pipeline configuration files.
+#
+# config.strings.support_escapes: false
```
Why not just `config.support_escapes: false` This applies to both `config.strings` i.e. `-e` as well as `-f` right?

jordansissel replied:
`strings` is in reference to string syntax in the grammar, not the "config string" (aka `-e`).

This is perhaps a bad name.

I wanted to imply that this was only for String grammar (https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/config/grammar.treetop#L75-L85) and nowhere else.

I'm open to calling it `config.support_escapes`

jordansissel replied:
Renamed to `config.support_escapes`

suyograo commented:
woot!

jordansissel commented:
Added specs:

```
LogStashConfigParser
  #compile
    when config.strings.support_escapes
      is enabled
        should process escape sequences
      is default
        should not process escape sequences
      is false
        should not process escape sequences

Finished in 0.26366 seconds (files took 2 seconds to load)
3 examples, 0 failures
```

jordansissel commented:
I also need to copy the AST change into lscl (https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/compiler/lscl.rb#L177-L181).

jordansissel commented:
I'll file an issue to get this change added to the LIR/LSCL treetop stuff and do it as a separate PR.

acchen97 commented:
@jordansissel you get a big heart from me for this one. :)

Just a minor note, do we want to roll this in 5.6, or wait for 6.0? If we introduce this in 6.0, we can directly add this by default as a breaking change and note it as a step in the major upgrade docs. It's been broken for a few years now, I'm not sure a few more months would inflict additional pain.

jordansissel commented:
The one test that failed looks like it stalled during an Elasticsearch output test:

https://travis-ci.org/elastic/logstash/builds/242544047?utm_source=github_status&utm_medium=notification

> No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

jordansissel commented:
@acchen97 From a version policy aspect, there's no reason it can't appear in 5.6 (default disabled) and 6.0 (default enabled, maybe, we can have that discussion). I think this setting being a boolean in logstash.yml gives us this flexibility to introduce it  during a minor (feature) release.

So, that's a long-winded way to say -- we can introduce this in a minor if desired because it's feature-flagged.

jordansissel commented:
@acchen97 I talked with @suyograo, and this should be good to land in 5.6 and 6.0.

After merging, I'll open a new issue to discuss enabling by default in 6.0.

andrewvc commented:
Is there a reason we wouldn't make the changes to lscl.rb in the same PR? I'm worried about diverging grammar here.

jordansissel commented:
@andrewvc in truth, I looked into this and my perception of the effort was that it would be much larger refactor and I don't feel I have the energy for it at this moment.

For this PR, I have chosen to pass the LogStash::Settings object to the compile() method. Maybe this wasn't the right choice, but the net impact, based on what I see, for lscl, is this:

`LogStashCompilerLSCLGrammar::Logstash::Compiler::LSCL::AST::String#expr` method needs to know if string escapes are enabled. If we do it the same way as this PR, this means the `expr` method needs to be given that information (currently, given the `LogStash::Settings` instance).

It is also that, to me, the LSCL grammar change can be done separate from this PR in the idea that I want to make smaller, not larger, PRs. The tests are passing, which means that we are not testing for parity between LSCL grammar/ast and the current grammar/ast.

andrewvc commented:
@jordansissel that's fine. We just need to make sure @ph and @original-brownbear are aware of this when they start implementing LIR execution. That update will be a blocker for #7291 . I've added it as a checkbox in that meta-issue.

jordansissel commented:
>  I'm worried about diverging grammar here.

I'm trying to balance the too-large-pr problem (adding LIR/LSCL refactor here) while trying to avoid divergence, so my thought was to recognize that more work is necessary and filed a bug to track it.

We could do this work in a feature branch, perhaps. I'm OK delaying this feature until LSCL support is added, if that is what you are asking for.

jsvd commented:
What do you think of making `support_escapes` a property of the `LogStash::Config::AST` that is set at startup time after reading the yaml file, or during [Config.compile](https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/config/config_ast.rb#L39-L43) ?

This way if we set the property during the first step of the compilation, like so:

```ruby
  class Config < Node
    def compile(settings)
      LogStash::Config::AST.deferred_conditionals = []
      LogStash::Config::AST.deferred_conditionals_index = 0
      LogStash::Config::AST.plugin_instance_index = 0
      LogStash::Config::AST.support_escapes = settings.get_value("config.support_escapes")
```

then we can avoid injecting the settings in all methods, as only the string node needs to grab its value:

```ruby
  class String < Value
    def compile(settings)
      if LogStash::Config::AST.support_escapes
        Unicode.wrap(LogStash::Config::StringEscape.process_escapes(text_value[1...-1]))
      else
        Unicode.wrap(text_value[1...-1])
      end
    end
  end
```

We already use LogStash::Config::AST as a hopping point for global state during compilation so I don't see a problem using the sigleton here too. This would drastically reduce the changes in this PR. What do you think?

jsvd commented:
I tested this locally, works wonderfully. I'm a bit concerned that people will abuse this, so I think all our documentation entries on this feature should have a complete list of things we escape. wdyt?

jordansissel commented:
> suggest that all our documentation entries on this feature should have a complete list of things we escape

+1

jordansissel commented:
Ahh, I didn't notice we used global state (something I was trying to avoid) already. I'll look into it -- I'd also like this PR to be smaller ;)

jordansissel commented:
Added a table in the docs with transform so now both references to string escapes are having a full list of what escapes are supported.

`bin/rspec logstash-core/spec` is passing for me

@jsvd  back to you for more review, I think.

jordansissel commented:
Oh, and I also  did the proposed refactor which makes this PR smaller by making the "should we process escape sequences?" checked as a property of the root AST node (Config).

Also has tests for both current grammar (config_ast) and future grammar (lscl)

jordansissel commented:
Partial test failures seem unrelated (time math, agent stop) to this PR.

jsvd commented on logstash-core/lib/logstash/environment.rb:
```diff
@@ -15,7 +15,7 @@ module Environment
   end
 
-  [
+  LOGSTASH_SETTINGS = [
```
I may be missing something but I don't see the need for this refactor

The `register_settings` method is only used in tests and right here, and even in the tests the support_escapes setting isn't done through a Settings object, but directly with the setter https://github.com/elastic/logstash/pull/7442/files#diff-1c97814b03bd1f3c0f796071ebc0cf19R169

Maybe this refactor isn't necessary?

jordansissel replied:
I may still use it in the specs, but I agree there are setters now for both config_ast and lscl. I'll try to revert the settings things.

jordansissel replied:
Ahha. mock_settings does what I need.

jordansissel commented:
@jsvd Ok I've removed my `register_settings` refactoring. I added it for a purpose that is already served (yay!) by another test feature (mock_settings).

jordansissel commented:
Finished in 5 minutes 53 seconds (files took 10.71 seconds to load)
1966 examples, 0 failures, 6 pending

Tests passing locally. I'll wait on travis and jenkins. @jsvd can you take another look?

jsvd commented on logstash-core/lib/logstash/compiler/treetop_monkeypatches.rb:
```diff
@@ -10,9 +10,10 @@ def set_meta(key, value)
   def compile
     return "" if elements.nil?
-    return elements.collect(&:compile).reject(&:empty?).join("")
+    return elements.collect { |x| x.compile }.reject(&:empty?).join("")
```
I see no need to keep this change and the white space above

jordansissel replied:
Ahh, yes. This was part of a refactor that is no longer needed.

jordansissel commented:
(Two test instances failed, but these failures were unrelated to this PR)

acchen97 commented:
WOOT! Excellent work! :)

andrewvc commented:
Very glad to see that land!

On Jun 21, 2017 1:57 AM, "Alvin Chen" <notifications@github.com> wrote:

> WOOT! Excellent work! :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/logstash/pull/7442#issuecomment-309981464>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIBYwXtQEqcHkS-an1NSam_V6xkAB_Vks5sGL7rgaJpZM4N430a>
> .
>

jordansissel commented:
5.x backport PR for this feature is here: https://github.com/elastic/logstash/pull/7516

binary132 commented:
Unicode escape sequences might be nice too (e.g. `\u00ff`).

jordansissel commented:
@binary132 Logstash pipeline configuration files are already expected to be UTF-8, so no escape sequences should be necessary since you can have these literally in your config:

```
input {
  file {
    path => "/var/log/*.log"
    add_tag => "☃︎"
  }
}
```

Hermain commented:
I'm using the docker gelf driver to send logs to logstash over tcp. The logs are NUL byte delimited (\0 or \u0000) unfortunately that still doesn't work. Its common in gelf to delimit messages with \0 and implementing this would remove the need for a gelf tcp plugin.
Is there a reason why \0 shouldn't be added to the supported list of escaped delimiters?

jordansissel commented:
hermain, I recommend you use the gelf input.

On Tue, Jan 9, 2018 at 6:51 AM Hermain <notifications@github.com> wrote:

> I'm using the docker gelf driver to send logs to logstash over tcp. The
> logs are NUL byte delimited (\0 or \u0000) unfortunately that still doesn't
> work. Its common in gelf to delimit messages with \0 and implementing this
> would remove the need for a gelf tcp plugin.
> Is there a reason why \0 shouldn't be added to the supported list of
> escaped delimiters?
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/logstash/pull/7442#issuecomment-356305429>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIC6l8iASs1W65ASUPZhVo59RaG_OgYks5tI3zTgaJpZM4N430a>
> .
>

Hermain commented:
But gelf input is udp only. If a small change (some way to use /0 as delimiter in a codec) is enough to make a plugin obsolete I would prefer that.
